ASK/TELL DEVELOP

jlnav opened the pull request:
See checklist in https://github.com/Libensemble/libensemble/pull/1307#pullrequestreview-2210671527

shuds13 commented on libensemble/tests/unit_tests/test_persistent_aposmm.py:
```diff
@@ -168,8 +168,88 @@ def test_standalone_persistent_aposmm_combined_func():
+    }
+
+    my_APOSMM = APOSMM(gen_specs)
+    my_APOSMM.setup()
```
Does this need to be separate to constructor?

jlnav replied:
Perhaps you've seen already, but `setup()` sets attributes that can't be pickled, but need to be done anyway. So I believe a separate `generator.setup()` needs to exist.

But perhaps for outside-of-libE purposes, to save a line, this could resemble: `my_APOSMM = APOSMM(gen_specs, setup=True)`.

shuds13 commented:
Currently this has two existing user functions refactored to ask/tell, and in a way that is a breaking change. I don't think this is what we want. So, for now, we could supply these as duplicates (keeping the original), or remove them. Alternative would be to refactor all appropriate gens to ask/tell, but that would hold this up.

jlnav commented:
> @jlnav
> 
> Currently this has two existing user functions refactored to ask/tell, and it is a breaking change for codes using those gens. I don't think this is what we want. So, for now, we could supply these as duplicates (keeping the original), or remove them. Alternative would be to refactor all appropriate gens to ask/tell, but that would hold this up, so I think it may be better to supply these two as a duplicate for now.

Sounds good. Maybe we could raise DeprecationWarnings?

> Also, would it be better for these ask/tell gens (rand sample and gpCAM) to use AskTellGenRunner and not use the wrapper gen_f?

I prefer the AskTellGenRunner myself (I _did_ develop it), but if all the gens move to my runner, then what would you want to do with your wrapper?

shuds13 commented:
I think there is also some opportunity for inheritence with the gpCAM class. And gp_cam_simple is more complicated than gp_cam_asktell. So, if anything the latter would be the base class, and renamed to avoid confusion.

shuds13 commented on libensemble/libE.py:
```diff
@@ -441,6 +441,24 @@ def libE_mpi_worker(libE_comm, sim_specs, gen_specs, libE_specs):
 # ==================== Local version ===============================
 
 
+def _retrieve_generator(gen_specs):
```
I don't know why this is needed. If gen is on a thread, should not need to be pickled.

shuds13 replied:
I've removed on my branch for now.

jlnav commented:
Checking random sample when running without Optimas:


```
>>> import numpy as np
>>> normal = np.load("persistent_aposmm_nlopt_history_length=2005_evals=2000_workers=4.npy")
>>> asktell = np.load("persistent_aposmm_nlopt_asktell_history_length=2002_evals=2000_workers=3.npy")
>>> all([i in asktell["x"][:100] for i in normal["x"][:100]])
True
```

jlnav commented:
With Optimas, currently looks like random sample doesn't match, even when trying to account for seed

jlnav commented:
Precision fixed in the Optimas example: https://github.com/optimas-org/optimas/pull/209/commits/08a835bb2d4efa30b76b9d86acf071c7d544310d

shuds13 commented:
We may need to update gpCAM gens for latest gpCAM release. Make sure any changes made in gen_f is reflected here before pulling in.

- [ ] Check gpCAM updates

shuds13 commented on libensemble/gen_classes/gpCAM.py:
```diff
@@ -0,0 +1,150 @@
+            self.all_y = np.vstack((self.all_y, self.y_new))
+
+            if self.my_gp is None:
+                self.my_gp = GP(self.all_x, self.all_y, noise_variances=self.noise * np.ones(len(self.all_y)))
```
I need to update for new gpCAM interface

shuds13 commented on libensemble/gen_classes/sampling.py:
```diff
@@ -0,0 +1,48 @@
+    """
+
+    def __init__(self, _, persis_info, gen_specs, libE_info=None) -> list:
+        # self.H = H
```
Remove I guess.

shuds13 commented on libensemble/gen_classes/gpCAM.py:
```diff
@@ -0,0 +1,150 @@
+        self.all_y = np.empty((0, 1))
+        np.random.seed(0)
+
+    def __init__(self, H, persis_info, gen_specs, libE_info=None):
```
I will put above _initialize_gpcAM

shuds13 replied:
and make  _initialize_gpcAM  _initialize_gpCAM

shuds13 commented on libensemble/gen_classes/gpCAM.py:
```diff
@@ -0,0 +1,150 @@
+        self.all_x = np.empty((0, self.n))
+        self.all_y = np.empty((0, 1))
+        np.random.seed(0)
+
```
We need to decide __init__ interface. 
We questioned before where we keep the same interface - which mirrors the current gen_f, or to rearrange, as H is often not given (basically an H0). So it could be gen_specs first. I'm leaning towards keeping the original ordering as it mirrors our user functions, but this should be discussed. 

We could also

jlnav replied:
Fair enough. My opinion/intuition is a user is more likely to prefer either "classical" gens (e.g. Jeff) or ask/tell gens (e.g. other CAMPA folks). With these gens' interfaces and users being so different, I don't think an arguably simpler rearrangement of the input parameters is too confusing.

Similarly to how some people prefer numpy or pandas; they do similar things, but their interfaces being different isn't a point of contention.

I'd also lean towards if someone were to initialize some object, like a gen, themselves, they'd prefer their specifications be provided as early and clearly as possible:

```
my_gen = Generator(param=1,
     option="two",
)
```

vs.

```
my_gen = Generator(None,
     {},
     {"param": 1, "option": "two"},
     {},
)
```

shuds13 reviewed (changes requested):
If we pull this in, are we treating this as provisional, without documenting gen_classes, so interface (e.g. `__init__` option ordering) is mutable. Or are we ready to say a user can use the gen_classes interface for their own gens if desired? This questions applies to the numpy interface (`ask_np`/`tell_np`). As for the outer API, we certainly have not finalized that, and it should somehow be noted.

E.g.,
```# This feature is in beta and its interface may change in future releases```

Note, that currently the two implemented classes (rearranged to ask/tell) are duplicates of gen_f, but in the long term we will not want to have duplicates.

Discussions:
- [x] Agree current status of gen_classes (internal or public).
- [x] If provisional - mark somehow in code
- [x] Agree wrapper paradigm (outer v inner interface)
- [x] Agree naming ask_np / tell_np
  - still not sure
  ```python
    ask_numpy()
   ```
   is better than 
   ```python
     @convert_dict_to_numpy
     ask()
   ```
  It seems we need separate function names, and the output of an ask() would be ambiguous.

- [x] Do we allow passing a class for ask/tell (if a libE generator) so initialize internally.
 - this is currently supported with the wrapper method but not internal ask/tell method
 - I think we should probably not support both wrapper and ask/tell approaches.
- [x] Agree num_points for ask policy (initial and thereafter):
 - Always empty ask and generator has to decide (most similar to now but what about external gens?)
 - intercept:
   -  init: `gen_specs['user']` fields `initial_batch_size` or `gen_batch_size` and else use num workers
   - after: intercept num points coming in for tell and ask for that many
   - Problem with intercepting `gen_specs['user']`  is names for batch size may be different.
 - Add fixed gen_specs options `gen_specs['initial_batch_size']`
   - and maybe something like `gen_specs['return_policy']` (or better name)
     - fixed batch size or generate num points got back or let generator decide points - empty ask()
     - (at some point may replace alloc_specs `async_return` if we want most user not to worry about allocs)?
 

- [x] Data conversion - how treat multi-dimensional fields.
 - currently makes a numpy array (as a dictionary value).
 
- [x] Agree gpCAM class names (e.g., GP_CAM_Covar) - and match gen_func names (original functions).
- [ ] Naming convention for tests - we have:
  - test_gpCAM_class.py
  - test_asktell_surmise.py
  - test_persistent_surmise_killsims_asktell.py
  - test_persistent_aposmm_nlopt_asktell.py
  - test_sampling_asktell_gen.py

- [x] Do our gens all support passing in intiail points? And if so, in standard format?

Interfacer:
- [ ] Does it work with executor (consideration of things added to executor in worker.py - including first-tier comm object)
- [x] Does it work when there is no executor
- [ ] For user function should we insert comm into libE_info["comm"] inside qcomm_main
- [x] Works with processes and threads - or picked one (and named right in generators.py)
- [x] Works with fork and spawn (for spawn be clear what in calling script needs to be inside main block).
- [x] Sometimes get hang at end of runs using the interfacer - esp. if no final_tell, is there something not getting shut-down? Check this works.
- [x] Need to deal with final_tell when using gen (e.g. optimas)
- [x] Need to sepaarate final_tell as required by standard - tell and extract data function.
- [ ] Constructor wrapper approach - demonstrate alternatives in notebook

Todo:
- [x] SH - update gpCAM to reflect current interface.
- [x] Change/rename original gpCAM to match naming in the class.
- [x] Need to run 'extra' tests for gpCAM tests to get run. Check passes.
- [x] Check correct no. points produced each call of gen (usually number sent back).

Check works:
- [x] Running libE ask/tell generator - e.g. gpCAM through Optimas (it will go via ask/tell wrappers).
- [x] Use libEnsemble ask/tell generator standalone (outside of libE)
 - must libE be installed to use generators?
- [x] Running libE with our ask_numpy/tell_numpy functions.
- [x] Running libE with ask/tell function (standardised format).

jmlarson1 commented on libensemble/generators.py:
```diff
@@ -0,0 +1,293 @@
+
+class Generator(ABC):
+    """
+    v 0.7.2.24
```
Is this version number something we need?

shuds13 commented on libensemble/generators.py:
```diff
@@ -0,0 +1,300 @@
+class LibensembleGenerator(Generator):
+    """Internal implementation of Generator interface for use with libEnsemble, or for those who
+    prefer numpy arrays. ``ask/tell`` methods communicate lists of dictionaries, like the standard.
+    ``ask_np/tell_np`` methods communicate numpy arrays containing the same data.
```
Check docstring names match functions (ask_numpy?)

shuds13 commented on libensemble/generators.py:
```diff
@@ -0,0 +1,300 @@
+    """
+
+    def __init__(
+        self, gen_specs: dict, History: npt.NDArray = [], persis_info: dict = {}, libE_info: dict = {}
```
Here we have a different constructor signature to those functions in gen_classes.

shuds13 commented:
In the full tests: https://github.com/Libensemble/libensemble/actions/runs/10256911231/job/28376940998#step:22:684

```
 ---Test 7: test_asktell_surmise.py starting with local on 4 processes 
/home/runner/miniconda3/envs/condaenv/lib/python3.11/site-packages/pydantic/_internal/_config.py:322: UserWarning: Valid config keys have changed in V2:
* 'orm_mode' has been renamed to 'from_attributes'
  warnings.warn(message, UserWarning)
Traceback (most recent call last):
  File "/home/runner/work/libensemble/libensemble/libensemble/tests/regression_tests/test_asktell_surmise.py", line 90, in <module>
    H_out, _a, _b = borehole(list_dicts_to_np(point), {}, sim_specs, {"H_rows": np.array([point["sim_id"]])})
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/libensemble/libensemble/libensemble/utils/misc.py", line 102, in list_dicts_to_np
    first = list_dicts[0]
            ~~~~~~~~~~^^^
KeyError: 0
Error: The operation was canceled.
```

shuds13 commented on libensemble/utils/runners.py:
```diff
@@ -84,3 +97,117 @@ def _result(self, calc_in: npt.NDArray, persis_info: dict, libE_info: dict) -> (
+class LibensembleGenRunner(AskTellGenRunner):
+    def _get_initial_ask(self, libE_info) -> npt.NDArray:
+        """Get initial batch from generator based on generator type"""
+        H_out = self.gen.ask_numpy(self.initial_batch or libE_info["batch_size"])
```
How are these batch sizes determined. How do we support fixed and/or dynamic sizes (e.g., producing as many points as has been returned)?

shuds13 replied:
2. I thought we said you call ask_numpy if exists, else call ask() - or is that dealt with by inheriting from Generator instead of LibensembleGenerator. Is there a test with a Generator in the standardised format using list of dictionary interface? We could add one in gen_classes sampling.

shuds13 commented on libensemble/utils/runners.py:
```diff
@@ -84,3 +97,117 @@ def _result(self, calc_in: npt.NDArray, persis_info: dict, libE_info: dict) -> (
+    def _loop_over_gen(self, tag, Work):
+        """Interact with ask/tell generator that *does not* contain a background thread"""
+        while tag not in [PERSIS_STOP, STOP_TAG]:
+            batch_size = self.batch or Work["libE_info"]["batch_size"]
```
How are these batch sizes determined. How do we support fixed and/or dynamic sizes (e.g., producing as many points as has been returned)? Same below.

jlnav commented on libensemble/generators.py:
```diff
@@ -0,0 +1,300 @@
+
+        gen_specs["gen_f"] = aposmm
+        if len(kwargs) > 0:
+            gen_specs["user"] = kwargs
```
Ah, its a weird looking choice that made the following interface possible on the Optimas side, and possibly the libE side too if I refactor the aposmm_asktell regression test:

```
aposmm = APOSMM(
    initial_sample_size=100,
    localopt_method="LN_BOBYQA",
    sample_points=np.round(known_minima, 1),
    rk_const=0.5 * ((gamma(1 + (n / 2)) * 5) ** (1 / n)) / sqrt(pi),
    xtol_abs=1e-6,
    ftol_abs=1e-6,
    dist_to_bound_multiple=0.5,
    max_active_runs=4,
    lb=np.array([var_1.lower_bound, var_2.lower_bound]),
    ub=np.array([var_1.upper_bound, var_2.upper_bound]),
)
```

shuds13 commented on libensemble/generators.py:
```diff
@@ -0,0 +1,300 @@
+    def tell_numpy(self, results: npt.NDArray) -> None:
+        pass
+
+    def ask(self, num_points: Optional[int] = 0) -> List[dict]:
```
Should these not be the definitions in Generator?

shuds13 commented on libensemble/generators.py:
```diff
@@ -0,0 +1,300 @@
+        """Request the next set of points to evaluate."""
+        return np_to_list_dicts(self.ask_numpy(num_points))
+
+    def tell(self, calc_in: List[dict]) -> None:
```
should calc_in be 'results' - same as in Generator?

shuds13 commented on libensemble/utils/runners.py:
```diff
@@ -84,3 +97,117 @@ def _result(self, calc_in: npt.NDArray, persis_info: dict, libE_info: dict) -> (
+                H_out = np.append(points, updates)
+            else:
+                H_out = points
+            tag, Work, H_in = self.ps.send_recv(H_out)
```
Dont we need a `np_to_list_dicts(H_out)` before this. 

Also, in the list of dictionaries data structure. If we convert a multi-dimensional x field, what type is the dictionary value supposed to be? Is it supposed to be a numpy array. I wonder this because the idea of the simple data structures was in part meant to avoid numpy.

shuds13 commented:
For APOSMM example, H is dumped before final_tell call, and so does not have local_min in dumped H. Also, looks like other fields might be a bit different between dumped H and the one returned from final_tell.

jlnav commented:
> For APOSMM example, H is dumped before final_tell call, and so does not have local_min in dumped H. Also, looks like other fields might be a bit different between dumped H and the one returned from final_tell.

Yeah; this will be true for Optimas until they implement `final_tell`. Until then Optimas currently has no way of knowing what points APOSMM has internally marked as minima (since `ask_updates` was rejected).

Since this branch does implement `final_tell` _and_ respects APOSMM's updates via marking points in H, **libEnsemble**'s H and APOSMM's H should be perfectly in sync.

shuds13 commented on libensemble/gen_classes/aposmm.py:
```diff
@@ -0,0 +1,123 @@
+            persis_info = add_unique_random_streams({}, 2, seed=4321)[1]
+        super().__init__(History, persis_info, gen_specs, libE_info, **kwargs)
+        if not self.persis_info.get("nworkers"):
+            self.persis_info["nworkers"] = gen_specs["user"]["max_active_runs"]  # ??????????
```
This is not nworkers. It could be less than nworkers.

jlnav replied:
Right. I'll find a better way to specify nworkers to aposmm

shuds13 replied:
If its not used until the main loop, can it be captured, or atleast checked later?

shuds13 commented on libensemble/gen_classes/aposmm.py:
```diff
@@ -0,0 +1,123 @@
+                self._last_ask = self._last_ask[~min_idxs]
+        if num_points > 0:  # we've been asked for a selection of the last ask
+            results = np.copy(
+                self._last_ask[self._ask_idx : self._ask_idx + num_points]
```
When we are in this situation (perhaps in error) where APOSMM does not give us enough points (e.g. the current ask is for 6, but APOSMM has only supplied one) then this gets filled up with empty points. This should be handled somehow. If it should never arise, then as a error condition.

shuds13 commented on libensemble/gen_classes/aposmm.py:
```diff
@@ -0,0 +1,123 @@
+    def _enough_initial_sample(self):
+        """We're typically happy with at least 90% of the initial sample, or we've already told the initial sample"""
+        return (
+            self._n_buffd_results >= self.gen_specs["user"]["initial_sample_size"] - 10
```
this line looks very dubious. Why is 10 a hardcoded value?

shuds13 commented on libensemble/gen_classes/aposmm.py:
```diff
@@ -0,0 +1,123 @@
+        self._n_total_results += len(results)
+
+        if not self._told_initial_sample and self._enough_initial_sample:
+            self._tell_buf = self._tell_buf[self._tell_buf["sim_id"] != 0]
```
Why getting rid of sim_id 0?

shuds13 commented on libensemble/gen_classes/aposmm.py:
```diff
@@ -0,0 +1,123 @@
+        self._told_initial_sample = False
+
+    def _slot_in_data(self, results):
+        """Slot in libE_calc_in and trial data into corresponding array fields. *Initial sample only!!*"""
```
this is what you want:
`self._tell_buf[self._n_buffd_results:self._n_buffd_results + len(results)] = results`

shuds13 commented on libensemble/generators.py:
```diff
@@ -0,0 +1,206 @@
+
+    def ask(self, num_points: Optional[int] = 0) -> List[dict]:
+        """Request the next set of points to evaluate."""
+        return np_to_list_dicts(self.ask_numpy(num_points))
```
I think we will want to convert from numpy types to regular python types. This could be done inside np_to_list_dicts or here e.g.

```python
    @staticmethod
    def convert_np_types(dict_list):
        np_types = (np.float64, np.int64, np.int32, np.float32)
        return [
            {key: (value.item() if isinstance(value, np_types) else value) for key, value in item.items()}
            for item in dict_list
        ]

    def ask(self, num_points: Optional[int] = 0) -> List[dict]:
        """Request the next set of points to evaluate."""
        vals = LibensembleGenerator.convert_np_types(np_to_list_dicts(self.ask_numpy(num_points)))
        return vals
```

shuds13 commented on libensemble/tests/regression_tests/test_asktell_aposmm_nlopt.py:
```diff
@@ -0,0 +1,101 @@
+
+    H, persis_info, _ = workflow.run()
+
+    # Perform the run
```
Move comment up?

shuds13 commented:
Consequences of removing the gen_f wrapper.
Orchestration,  in terms of how you interface with a model, happens inside a gen_f.
This includes things like where we decide resources for each simulation.
For example, in the Optimus gen_f, it has an option to set `num_procs` or `num_gpus` for each simulation. This could be used with any generator . If we get rid of the gen_f wrapper where does this happen?

shuds13 commented on libensemble/gen_classes/aposmm.py:
```diff
@@ -0,0 +1,125 @@
+        persis_info: dict = {},
+        gen_specs: dict = {},
+        libE_info: dict = {},
+        **kwargs,
```
For libEnsemble arguments this is opaque when looking at this routine. Perhaps they should be written out as additional arguments. Then there's the question of whether you keep gen_specs as well.

jlnav committed:
add parameter for UniformSample, instructing the generator to squash all of VOCS.variables down to single "x"

jlnav committed:
making variables_mapping an explicit kwarg for a libensemble generator

jlnav committed:
bump pydantic - this should've been done in a previous PR

jlnav committed:
fix UniformSample LibensembleGenerator to use variables_mapping

jlnav committed:
just commenting out, but maybe we don't need this unit test now that the gen is tested in a regression test?

jlnav committed:
removing unnecessary GPCAM_Standard class; refactoring the other gpcam classes to accept VOCS

jlnav committed:
update regression test to (presumably) use classes that now use VOCS

jlnav committed:
... why was this test excluded?

jlnav committed:
lists instead of tuples...

jlnav committed:
fix default value of test_points_file kwarg to be None

jlnav committed:
somehow this tell_numpy got missed

jlnav committed:
various fixes to building variables_mapping and extra.ci, still debugging len(np.unique(H["gen_ended_time"])) != num_batches; it exceeds it

jlnav committed:
temporary debugging print?

jlnav committed:
fix objective name

jlnav committed:
batch_size fix to runners.py, add batch_size to gen_specs so the runners .ask(15) instead of the default 4, re-add the gpcam gen

jlnav committed:
we still want to map {"f": ["energy"]), obviously

jlnav committed:
move StandardSample to test_asktell_sampling.py, to test as-though we've imported an external generator

jlnav committed:
various fixes from PR suggestions

jlnav committed:
move gpcam superclass init to top of init, make UniformSample accept random seed

jlnav committed:
gpcam asserts objectives, then a better description of variables_mapping

jlnav committed:
move _validate_vocs call to superclass

jlnav committed:
Merge pull request #1572 from Libensemble/asktell/VOCS

Experiments with VOCS in our gen classes

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

shuds13 committed:
Split finalize and export

shuds13 committed:
aposmm uses x mapping to set bounds and size

jlnav committed:
starting to populate the APOSMM class with common kwargs, for better documenting, and so we don't have to check the existence of settings in kwargso

jlnav committed:
small fixes

jlnav committed:
docstring for APOSMM class

shuds13 committed:
Fix finalize and export functions

shuds13 committed:
Option to export with user fields

jlnav committed:
evaluate an APOSMM with only VOCS passed in

shuds13 committed:
Add unit tests of unmap_numpy_array

shuds13 committed:
Remove unneeded branch

shuds13 committed:
Add expected variables mapping for APOSMM

shuds13 committed:
Better example bounds

shuds13 committed:
Allow pass through of unmapped arrays

shuds13 committed:
Allow export as list of dictionaries

shuds13 committed:
Add pass-through array to unmap test

shuds13 committed:
Add export unit tests and fix up unmap

shuds13 committed:
Re-enable APOSMM unit tests

shuds13 committed:
Add checks for x and x_on_cube

shuds13 committed:
Add export tests and fixup

jlnav committed:
replace completely-typed out gen_specs['user'] update from parameters with loop over fields and grabbing the value from locals, as suggested by shuds

jlnav committed:
coverage adjusts

shuds13 committed:
Do not send local_min/pt to ingest

shuds13 committed:
Autofill x and f variables_mapping separately

shuds13 committed:
Update asktell APOSMM regression test

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

shuds13 committed:
Add fvec when components is present

shuds13 committed:
Send APOSMM errors as a string

shuds13 committed:
Formatting

shuds13 committed:
Blacken

shuds13 committed:
Clarify comment

shuds13 committed:
Update generator_standard to gest_api

shuds13 committed:
Fix gest-api in pyproject

shuds13 committed:
Fix gest project name

shuds13 committed:
Fix _validate_vocs for gpCAM

shuds13 committed:
Remove misleading n

shuds13 committed:
Merge pull request #1582 from Libensemble/feature/update_finalize

Update finalize to meet standard

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/aposmm_fixes

jlnav committed:
Merge branch 'develop' into asktell/aposmm_fixes

jlnav committed:
this specific gen_specs['out'] assignment no longer needed, as conflicts with using variables_mapping later on

jlnav committed:
adjust for finalize and export; plus now variables_mapping is required

jlnav committed:
don't need persis_info declared like this

jlnav committed:
fixes and refactors as suggested by shuds - *works on multistage lpa*

jlnav committed:
make initial_sample_size and max_active_runs required arguments for aposmm class - and rearrange to be higher in the docstring and signature

jlnav committed:
Merge pull request #1583 from Libensemble/asktell/aposmm_fixes

Better parameterization of APOSMM class, plus docstring

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

jlnav committed:
typo

jlnav committed:
fix generator-standard package name in pyproject.toml

jlnav committed:
also start running thread upon initial ingest

jlnav committed:
better docstrings. new "do_not_produce_sample_points" arg from shuds. an attempt to make the arg perform an extra receive so the user can pass in those values. add the boilerplate for an extra test

jlnav committed:
add kwarg where necessary, plus more test glue

jlnav committed:
unpack "x"s passed in into dtype-less array. so _on_cube conversion can happen as expected

jlnav committed:
TODOs, plus more

jlnav committed:
mostly just trying the current aposmm.py with Xopt in this commit...

jlnav committed:
fix initial sample (within aposmm *class*) requiring sim_ids. we'll just pass the data as-is into the gen_f

jlnav committed:
aposmm gen_f : new loop undserneath do_not_produce_sample_points condition where aposmm receives until enough have been received, then local_H is updated with the method used when libE has returned points. Then fix something_sent condition. Also the update_local_H_after_receiving resizes local_H in this initialization routine. Then in generators.py, ensure additionally that Work can ingest points without sim_id needing to be in results. Assume that an np.arange is good enough

jlnav committed:
add unit test where APOSMM class ingests first

jlnav committed:
make an internal method internal. implement _validate_vocs for aposmm, primarily displaying a series of warnings about the characteristics of vocs

jlnav committed:
trying to account for, then test a large variety of conditions where interacting with the aposmm class may fail. add supported globus_compute_sdk version to dev env

jlnav committed:
adjusts send-conditions upon full sample being ingested. adjusts warning message. appends _id to sim_id mapping in generators.py and removes the associated pop in misc.py

jlnav committed:
additional docs

shuds13 committed:
Add Xopt EI test

shuds13 committed:
Restructure imports

shuds13 committed:
Set batch size and fix nworkers

shuds13 committed:
Make temp note on fixing nworkers

shuds13 committed:
Restructure gen_specs

shuds13 committed:
Change sim to xopt test sim

shuds13 committed:
Add vocs field to GenSpecs

shuds13 committed:
Add vocs field to SimSpecs

shuds13 committed:
Get vocs field type

jlnav committed:
tentative refactor of do_not_produce_sample_points keyword to be the inverse

shuds13 committed:
Add unit tests for set_fields_from_vocs

shuds13 committed:
Test array type

jlnav committed:
doing "generate_sample_points". ensuring sim_id gets cast as int (for using as indexes later). need to accomodate circumstance where we have set up a sim_id <- _id mapping, but no input _id data is appearing, since this is acceptable for an initial sample

jlnav committed:
add onto n_s with the len of received presumptive_user_sample

shuds13 committed:
Prevent gen from converting to dictionary

shuds13 committed:
Cleanup test_xopt_EI.py

jlnav committed:
fix conditions for something_sent after initial_sample is ingested. adjust tests to match expected behavior

shuds13 committed:
Provide pre-evaluated initial sample

jlnav committed:
internally, generate_sample_points is now set based on whether suggest or ingest is called first

shuds13 committed:
Remove automapping

jlnav committed:
various fixes; sim_id warning no longer needed in gen_f. raise RunTimeError on unexpected usage instead of failing/returning silently. Additional tests for coverage

jlnav committed:
always map _id to sim_id. add simple executor + UniformSample test. clarifying comments

jlnav committed:
put the executor.comm adjustment back - and add a asktell aposmm + executor test

jlnav committed:
other MPI processes attempting to finalize their gen is problematic

shuds13 committed:
Add array dtype support

shuds13 committed:
Update vocs sampling tests

* Split tests libEnsemble and external generators
* Have an vocs generator that uses arrays
* Array generator disabled as requires gest-api update.

shuds13 committed:
Remove redundant line

shuds13 committed:
Remove another redundant line

shuds13 committed:
Formatting

shuds13 committed:
Add gest_api simulator wrapper

shuds13 committed:
Add version of xopt test that uses xopt simulator

shuds13 committed:
Fix naming

shuds13 committed:
Fix test for array fields

shuds13 committed:
Disable awkward array test

shuds13 committed:
Give output assert to xopt tests

shuds13 committed:
Fix formatting

shuds13 committed:
Merge pull request #1622 from Libensemble/examples/xopt_generators_remove_auto_mapping

Remove automapping

shuds13 committed:
Add Optimas grid sample test

shuds13 committed:
Infer type of discrete vars

jlnav committed:
tweaks based on PR feedback

shuds13 committed:
Re-enamble model tests

shuds13 committed:
Add xopt sequential test

shuds13 committed:
Do not ingest None

shuds13 committed:
Simplify xopt nelder mead test and assert

shuds13 committed:
Correct procs count for when use parse_args

shuds13 committed:
Add xopt and optimas to extra CI

shuds13 committed:
Formatting

shuds13 committed:
Fix install

shuds13 committed:
Test xopt/optimas gens on basic tests

jlnav committed:
doc/docstring additional tweaks and clarifications

jlnav committed:
bump polling_loop interval for (perhaps) slower containers/VMs/systems (like the macOS jobs)

jlnav committed:
set the sim_ids of the ingest_buf right when we're about to send it to the gen_f. add a initialize_dists_and_inds call upon the initial sample being ingested in, inside update_local_H_after_receiving

shuds13 committed:
Fixing CI for xopt/optimas generator tests (#1626)

* Install CPU-only pytorch

* Xopt install with no updating of deps

* Install Optimas from github main

* Temporarily only run target tests.

shuds13 committed:
Clean up

shuds13 committed:
Add xopt notebook

shuds13 committed:
Add xopt example to docs

shuds13 committed:
Add new tutorial to main toctree

shuds13 committed:
Rename example

shuds13 committed:
Rename and add colab link

shuds13 committed:
Improve layout

shuds13 committed:
Add xopt/forces example

shuds13 committed:
Enable gest-api simulator to use executor

shuds13 committed:
Fix user_specs naming

shuds13 committed:
Formatting

shuds13 committed:
Merge pull request #1620 from Libensemble/examples/xopt_generators

Add Xopt and Optimas generator tests

jlnav committed:
Merge branch 'develop' into asktell/nov_aposmm_refactor

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/nov_aposmm_refactor

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/nov_aposmm_refactor

shuds13 committed:
Merge pull request #1616 from Libensemble/asktell/nov_aposmm_refactor

APOSMM accepts an initial sample, via ingest

shuds13 committed:
Ensure gens with _id return with sim_id

shuds13 committed:
Add Optimas Ax test

shuds13 committed:
Add working multi-task test

shuds13 committed:
Target Optimas branch for testing

shuds13 committed:
Add tests with single/multi fidelity and multitask gens

shuds13 committed:
Remove old optimas test

shuds13 committed:
Make _id to sim_id more robust

shuds13 committed:
Handle empty dictionary

shuds13 committed:
Formatting tests

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

jlnav committed:
typo

shuds13 committed:
Ingest initial H0 for gest-api generators

shuds13 committed:
Multitask test uses H0

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

shuds13 committed:
Fix test naming

jlnav committed:
initial small changes to add returns_id attribute to APOSMM class

jlnav committed:
adjust README intro sample to use standard gen, VOCS

jlnav committed:
de-emphasize the "user-functions" in favor of generators and simulators. separate out the allocator discussion

jlnav committed:
Merge pull request #1646 from Libensemble/asktell/aposmm_id_attr

initial small changes to add returns_id attribute to APOSMM class

jlnav committed:
mention uv as an install approach. in README example vocs->variables_objectives for clarity. remove libE() doc

jlnav committed:
remove sim_specs / gen_specs docs components that describe defining those structures as dictionaries

jlnav committed:
bump specifically-mentioned required pydantic version in advanced_installation

jlnav committed:
initial files and fixes for autodocing the gest-api adhering generators

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' of https://github.com/Libensemble/libensemble into experimental/jlnav_plus_shuds_asktell

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/emphasize_asktell_gens

jlnav committed:
small fixes, plus display some ask/tell aposmm examples in the docs

jlnav committed:
list gpcam example

jlnav committed:
monocasing, plus a clarification

jlnav committed:
linenos

jlnav committed:
starting to rewrite the first tutorial for standardized gens

jlnav committed:
remove some redundancy

jlnav committed:
link to xopt/optimas generators in docs - also try only doc'ing suggest/ingest/export/finalize for our generators instead of suggest/ingest_numpy

jlnav committed:
note the third party gens are examples - adjust xopt file paths to main branch

jlnav committed:
add gest-api pypi release as required dependency

jlnav committed:
Merge branch 'develop' into asktell/gest-api-release

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

jlnav committed:
lock scipy version until pixi changes implemented into extra.yml

jlnav committed:
lock to version from a bit ago

jlnav committed:
remove lock...?

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/emphasize_asktell_gens

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/gest-api-release

jlnav committed:
fix psutil version on macos in lockfile

jlnav committed:
Merge pull request #1666 from Libensemble/asktell/gest-api-release

add gest-api pypi release as required dependency

jlnav committed:
unbump ax

jlnav committed:
Merge branch 'develop' into asktell/emphasize_asktell_gens

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/emphasize_asktell_gens

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

shuds13 committed:
Add _id field when id producing gen

shuds13 committed:
Handle _id as own field

shuds13 committed:
Make test_optimas_ax_multitask.py local only

shuds13 committed:
Merge from upstream

shuds13 committed:
Formatting update

shuds13 committed:
Formatting

shuds13 committed:
Update xopt branch

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

shuds13 committed:
Put back _id to sim_id fixup

shuds13 committed:
Merge from upstream

jlnav committed:
fix generator.vocs attribute to be lowercase throughout. fix bugged lockfile url to gest-api package

shuds13 committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into examples/optimas_ax_generators

jlnav committed:
add optimas to all extra envs. add xopt to py3.11+ envs

jlnav committed:
since xopt and optimas now included in envs, remove xopt tests on python 3.10

jlnav committed:
use main branch of optimas instead of release

jlnav committed:
forgot an e

jlnav committed:
grab develop's run-tests.py

jlnav committed:
why are these tests complaining about reusing an ensemble dir

jlnav committed:
pass the instance, not the parent class that was included for typing purposes

shuds13 committed:
Exclude multitask

jlnav committed:
enabled / disabled tests adjustments

jlnav committed:
move nlopt dep

jlnav committed:
upstream optimas branch for asktell

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into examples/optimas_ax_generators

jlnav committed:
some fixes (?), experimenting with calling only suggest on LibensembleGenInterfacer's, plus refactoring the aposmm calling script into how I'd like it to look eventually.

jlnav committed:
Merge remote-tracking branch 'remotes/origin/examples/optimas_ax_generators' into asktell/aposmm_reg_test_vocs

jlnav committed:
various fixes, adjustments, new functions. unmap aposmm's output numpy array into the H constructed from VOCS, then do the reverse for the numpy array from H back into APOSMM. Also ensure that _id data becomes sim_id, and vice-versa

jlnav committed:
history.py:
- Hint towards easy bug where user may have intended to set sim_specs.simulator instead of sim_specs.sim_f

the regression test:
- use vocs for second test iteration

misc.py
- more reliably build the dtype out of mapping keys that are present in the input data

runners.py
- some cleanup

jlnav committed:
exclude optimas ax tests on macos

jlnav committed:
remove optimas-ax tests from newer python extended tests

jlnav committed:
Merge pull request #1635 from Libensemble/examples/optimas_ax_generators

Examples/optimas ax generators

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/aposmm_reg_test_vocs

jlnav committed:
Merge branch 'develop' into experimental/jlnav_plus_shuds_asktell

jlnav committed:
coverage fixes, plus adjustments for warning/handling when both the wrong simulator/sim_f parameter is set in sim_specs, and accordingly the wrong data is returned. test for this

jlnav committed:
adjusts for MPI case

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/aposmm_reg_test_vocs

jlnav committed:
dependency and expected-error-condition tweaks

jlnav committed:
lock to previous petsc version

shuds13 committed:
Merge pull request #1679 from Libensemble/asktell/aposmm_reg_test_vocs

Asktell/aposmm reg test with vocs + other changes

jlnav committed:
Merge branch 'experimental/jlnav_plus_shuds_asktell' into asktell/emphasize_asktell_gens

jlnav committed:
Merge pull request #1656 from Libensemble/asktell/emphasize_asktell_gens

Docs for ask/tell gens

jlnav committed:
open up ax-platform across extra-envs for compatiblity with xopt requirements

jlnav committed:
fix tutorial code

jlnav committed:
typo

jlnav committed:
lock ax again, point to optimas branch

jlnav committed:
disable xopt install/tests for now

jlnav committed:
Merge pull request #1684 from Libensemble/asktell/ax

Asktell/ax - disable xopt tests for now and narrow down some deps
